src/sshfs_mount/sshfs_mount.cpp

```diff
@@ -62,6 +81,43 @@ void check_sshfs_exists(mp::SSHSession& session)
     run_cmd(session, "which sshfs", error_handler);
 }
 
+// If the target name starts with ~, this function returns the full directory name. If not, it returns the same
+// target name.
+std::string expand_home_directory(mp::SSHSession& session, const std::string& target)
+{
+    std::string expanded_target;
+
+    if ('~' == target[0])
+    {
+        std::string home;
+
+        // Get the user name, which can be empty.
+        auto pos = (1 == target.size()) ? 1 : target.find('/', 1);
+        if (1 == pos)
+        {
+            home = run_string_cmd(session, "pwd");
+        }
+        else
+        {
+            std::string username = target.substr(1, pos - 1);
+            home = run_string_cmd(session, "getent passwd " + username + " | cut -d : -f 6");
```

Comment:
Just realized there's a problem with this - `getent passwd 1000` works, too, but I'm unsure if `multipass mount . instance:~1000` should work (I mean, getent works with UIDs, not only usernames, but `~` in a shell does not).

Reply:
Looking at [`man adduser.conf`](http://manpages.ubuntu.com/manpages/trusty/man5/adduser.conf.5.html), usernames by default need to begin with a `[a-z]`.

Maybe even more importantly, in [`man getent`](http://manpages.ubuntu.com/manpages/xenial/en/man1/getent.1.html):
> pass each numeric key to getpwuid(3) and each nonnumeric key to getpwnam(3)

So there's three routes that I can see:
1. ignore this, and support UIDs as well as usernames, and let `getent` deal with it
1. only accept non-numeric usernames and treat e.g. "~1000" verbatim
1. accept the "unsafe" option of letting the shell evaluate the whole string

Only other option would be to somehow call [`getpwnam()`](http://manpages.ubuntu.com/manpages/eoan/en/man3/getpwnam.3.html) explicitly inside the instance…

Reply:
I didn't think about numeric ID's.

My thought here is that it is safe to use option 1. We can make some check before calling `getent`, to avoid wrong strings. Then `getent` deals with it, and errors back if the user does not exist. Option 2 is not good I think, because using option 1 the user can still mount on a path containing `~`. Option 3 sounds to me unsafer than option 1.

Reply:
The problem is that if you _do_ have a user "1000:1001:1001…" in `/etc/passwd` (just add manually), that will not be selected over the user with UID 1000…

Reply:
Oh, right, we were discussing corner cases :P
So we don't have a best option. In this case, 3 would be the best because the shell will resolve things that we can't do without much job. But at the cost of loosing a bit of safety, we should perform some checks there. But then, again, we'll have to make some assumptions on the user name, i.e., it cannot contain ';' or something used for injecting commands.

Reply:
Wait, thinking again here. We should use option 3. Multipass should behave as the shell would do. If we do `sudo adduser 1000 --force-badname`, then `getent passwd 1000` would yield `ubuntu:x:1000:1000:Ubuntu:/home/ubuntu:/bin/bash` but `~1000` will evaluate to `/home/1000`. I think the user would expect Multipass to behave like the shell, no matter what getent does. We definitely should let the shell evaluate, and this would imply forgetting about numerical IDs for mounting.

Reply:
Yeah, the argument against that is the "it's unsafe" consideration. But as I said before, they have `root` on the instance already…